forge-ai/src/main/java/forge/ai/ComputerUtil.java

```diff
@@ -1165,42 +1177,244 @@ public static boolean castPermanentInMain1(final Player ai, final SpellAbility s
             }
         }
 
-        // cast Backup creatures in main 1 to pump attackers
-        if (cardState.hasKeyword(Keyword.BACKUP)) {
-            for (Card potentialAtkr: ai.getCreaturesInPlay()) {
-                if (ComputerUtilCard.doesCreatureAttackAI(ai, potentialAtkr)) {
+        // Triggered abilities
+        for (Trigger trigger : cardState.getTriggers()) {
+            final String triggerZones = trigger.getParam("TriggerZones");
+            if (!"Battlefield".equals(triggerZones)) continue;
+
+            final TriggerType mode = trigger.getMode();
+            if (mode != null) {
+                // 1312+ cards match `Mode\$ ChangesZone.*?TriggerZones\$ Battlefield`
+                // enter/leave triggers
+                // Admonition Angel etc
+                if (mode == TriggerType.ChangesZone || mode == TriggerType.ChangesZoneAll) {
+                    return true;
+                }
+
+                // 1009+ cards match `Mode\$ SpellCast.*?ValidActivatingPlayer\$ You`
+                // Aetherflux Reservoir etc
+                if (mode == TriggerType.SpellCast && "You".equals(trigger.getParam("ValidActivatingPlayer"))) {
```

Comment:
sequence wise this function should decide if there is something important enough to cast before combat (vs. leaving mana open for combat tricks)
_Aetherflux Reservoir_ does not qualify, it should just be cast before other spells ideally so that's more of a use case for `AIPriorityModifier`